### PR TITLE
Added `hclvalidate` command

### PR DESCRIPTION
**Pull request by levkohimins**
## Description

The PR is adding a new CLI command that searches for all HCL files in the config stack and validates them.

https://linear.app/gruntwork/issue/TG-19/improve-pre-flight-validation-tooling-in-terragrunt#comment-73a32adb

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [ ] Update the docs.
- [ ] Run the relevant tests successfully, including pre-commit checks.
- [ ] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [ ] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

Added / Removed / Updated [X].

### CODE REVIEW

File: config/locals.go [hunk 1]
```diff
@@ -67,12 +68,18 @@ func evaluateLocalsBlock(ctx *ParsingContext, file *hclparse.File) (map[string]c
...
+		ctx.TerragruntOptions.Logger.Debugf("Not all locals could be evaluated:")
+		var errs *multierror.Error
+		for _, attr := range attrs {
+			diags := canEvaluateLocals(attr.Expr, evaluatedLocals)
+			if err := file.HandleDiagnostics(diags); err != nil {
+				errs = multierror.Append(errs, err)
+			}
+		}
+
+		if err := errs.ErrorOrNil(); err != nil {
```
[the remark this answers is not in the document]
Comment (levkohimins) [reply config/locals.go:80]: Yes, but it won't cause panic.

https://github.com/hashicorp/go-multierror/blob/6fa16eae919d27fdd2a107373fff9388795a6802/multierror.go#L31-L33

File: configstack/module.go [hunk 2]
```diff
@@ -47,115 +47,343 @@ func (module *TerraformModule) String() string {
...
-		result, err := flagModulesThatDontInclude(includedModulesWithExcluded, terragruntOptions)
+	crossLinkedModules, err := runningModules.crossLinkDependencies(dependencyOrder)
+	if err != nil {
+		return crossLinkedModules, err
+	}
+
+	return crossLinkedModules.removeFlagExcluded(), nil
+}
+
+// Check for dependency cycles in the given list of modules and return an error if one is found
```
[the remark this answers is not in the document]
Comment (levkohimins) [reply configstack/module.go:361]: Fixed

File: configstack/errors.go [hunk 3]
```diff
@@ -32,3 +41,41 @@ type InfiniteRecursion struct {
...
+	Module     *TerraformModule
+	Dependency *TerraformModule
+	Err        error
+}
+
+func (err DependencyFinishedWithError) Error() string {
+	return fmt.Sprintf("Cannot process module %s because one of its dependencies, %s, finished with an error: %s", err.Module, err.Dependency, err.Err)
+}
+
+func (this DependencyFinishedWithError) ExitStatus() (int, error) {
```
[the remark this answers is not in the document]
Comment (levkohimins) [reply configstack/errors.go:63]: Actually this is not my code, I only moved it when refactoring the `configstack` package.

https://github.com/gruntwork-io/terragrunt/blob/7699c22743b4d5a11314fdfbd8ab69fadab057b9/configstack/running_module.go#L311-L322

[the remark this answers is not in the document]
Comment (levkohimins) [reply configstack/errors.go:63]: I have renamed to `ErrorProcessingModuleDependency`.

[the remark this answers is not in the document]
Comment (levkohimins) [reply configstack/errors.go]: Fixed

File: test/integration_test.go [hunk 4]
```diff
@@ -213,6 +213,115 @@ const (
 	fixtureDownload = "fixture-download"
 )
 
+func TestHclvalidateDiagnostic(t *testing.T) {
+	cleanupTerraformFolder(t, TEST_FIXTURE_HCLVALIDATE)
```
Comment (denis256) [inline test/integration_test.go:217]: no `t.Parallel()` - it is by intention?

File: internal/view/human_render.go [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+	default:
+		// Clear out any coloring that might be applied by Terraform's UI helper,
+		// so our result is not context-sensitive.
+		buf.WriteString(render.colorize.Color("\n[reset]"))
+	}
+
+	// We don't wrap the summary, since we expect it to be terse, and since
+	// this is where we put the text of a native Go error it may not always
+	// be pure text that lends itself well to word-wrapping.
+	fmt.Fprintf(&buf, render.colorize.Color("[bold]%s[reset]\n\n"), diag.Summary)
```
Comment (denis256) [inline internal/view/human_render.go:104]: looks like error returned from `Fprintf` is not handled

File: internal/view/diagnostic/function.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+func DescribeFunctionParam(p *function.Parameter) FunctionParam {
+	ret := FunctionParam{
+		Name: p.Name,
+	}
+
+	if raw, err := p.Type.MarshalJSON(); err != nil {
+		// Treat any errors as if the function is dynamically typed because it would be weird to get here.
+		ret.Type = json.RawMessage(`"dynamic"`)
+	} else {
+		ret.Type = json.RawMessage(raw)
```
Comment (denis256) [inline internal/view/diagnostic/function.go:34]: Seems to be redundant type conversion  (same in other places too)

File: internal/view/diagnostic/expression_value.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+	"github.com/hashicorp/hcl/v2"
+	"github.com/zclconf/go-cty/cty"
+)
+
+const (
+	// Sensitive indicates that this value is marked as sensitive in the context of Terraform.
+	Sensitive = valueMark("Sensitive")
+
+	// TypeType is used to indicate that the value contains a representation of another value's type.
+	TypeType = valueMark("TypeType")
```
Comment (denis256) [inline internal/view/diagnostic/expression_value.go:17]: TypeType - seems to be unused

File: test/integration_test.go (hunk 4 above)
Comment (levkohimins) [reply test/integration_test.go:217]: Ah, yes, thanks. Fixed.

File: internal/view/human_render.go (hunk 5 above)
Comment (levkohimins) [reply internal/view/human_render.go:104]: Fixed

File: internal/view/diagnostic/function.go (hunk 6 above)
Comment (levkohimins) [reply internal/view/diagnostic/function.go:34]: Fixed

File: internal/view/diagnostic/expression_value.go (hunk 7 above)
Comment (levkohimins) [reply internal/view/diagnostic/expression_value.go:17]: Fixed

Comment (ZachGoldberg) [reply internal/view/diagnostic/expression_value.go]: test comment please ignore

**Comment by ZachGoldberg** [comment]
test comment please ignore #2

File: internal/view/diagnostic/expression_value.go (hunk 7 above)
Comment (ZachGoldberg) [reply internal/view/diagnostic/expression_value.go]: test comment num 3

**Comment by ZachGoldberg** [comment]
test comment 4

**Comment by ZachGoldberg** [comment]
> test comment 4

Number 5

**Review by denis256 — APPROVED** [review]

**Comment by levkohimins** [comment]
@denis256 thanks for the review!

**Pull request merged**